#164 Implement new GameCursor class

mwnDK1402 opened the pull request:
- [x] Fix cursor being visible when tabbing back into the game
- [x] Save and load in-game camera mode in PlayerPrefs
- [x] Replace calls to `Cursor` and `Screen.lockCursor` with calls to `GameCursor`
- [x] Enter Loading mode when a level is loading

Closes #164

JupiterJaeden committed:
First attempt at CursorManagement class

It doesn't work properly, but I'll try and fix it later.

JupiterJaeden committed:
Read the description

Doesn't fix the issue, but I just want to get this committed because it's a start.

mwnDK1402 committed:
Rewrite CursorManagement

Blame Alex for mistakes

mwnDK1402 committed:
Replace old calls to Cursor with CursorManagement

mwnDK1402 committed:
Change menu key from Esc to P

mwnDK1402 commented:
We should probably proceed in this order:
- Merge `development` into `164-cursor-overhaul`
- Use Loading mode when loading
- Clean the code

When it comes to cleaning the code, I wrote some TODO comments, mentioning places I thought might need some more investigation and testing.

mwnDK1402 commented on Assets/Scripts/UI/CursorManagement.cs:
```diff
@@ -0,0 +1,102 @@
+    private static Mode PreferredMode = Mode.Original;
+    private static Mode CurrentMode = Mode.Menu;
+
+    public enum Mode : int
```
@Lithrun prefers not specifying the underlying type, and he wants public enums to have their own file.

mwnDK1402 committed:
Resolve merge conflicts

mwnDK1402 committed:
Fix errors and broken PlayerTitan hair

mwnDK1402 committed:
Switch to Loading mode when creating room and add menu toggled event

mwnDK1402 committed:
Set default for AttackAnimation in JumpAttack

JupiterJaeden committed:
Update Hero.cs

mwnDK1402 committed:
Refactor CursorManagement -> GameCursor

Refactoring involves keeping track of open menus.

mwnDK1402 committed:
Resolve conflict by keeping local changes

mwnDK1402 committed:
Merge with development

mwnDK1402 commented on Assets/Scripts/Characters/Titan/Attacks/GrabAttack.cs:
```diff
@@ -8,7 +8,9 @@ public GrabAttack()
         }
-        private string AttackAnimation { get; set; }
+
+        // HACK: It is Empty in certain situations. This fixes that.
```
Let's find out what issue should fix this.

Lithrun replied:
Create a new bug with reproduction steps.

mwnDK1402 commented on Assets/Scripts/Characters/Titan/PlayerTitan.cs:
```diff
@@ -20,6 +20,12 @@ public class PlayerTitan : MindlessTitan
+            // TODO: Make the other camera modes work.
+            // This also unfortunately sets the preferred
+            // camera mode. But that will change when we fix it.
+            GameCursor.CameraMode = CameraMode.TPS;
```
Try commenting out this line and see what happens. I'm not even sure what the intended behaviour is for player titans. I don't think it makes sense to actually have the normal camera modes apply to titans, but setting the camera mode like this sets the preferred camera mode, which messes with the player's settings...

Lithrun replied:
Player Titans can have TPS, Original and WOW camera. Don't see a reason why this functionality should be removed

mwnDK1402 commented on Assets/Scripts/Characters/Titan/TITAN_SETUP.cs:
```diff
@@ -15,7 +15,7 @@ public class TITAN_SETUP : Photon.MonoBehaviour
     private void Awake()
     {
-        this.hair_go_ref = new GameObject();
+        this.hair_go_ref = new GameObject("Hair Reference");
```
I just gave it a name so it was easier to find in the scene.

mwnDK1402 commented on Assets/Scripts/FengGameManagerMKII.cs:
```diff
@@ -863,23 +859,25 @@ private void coreeditor()
+                // TODO: Mouselook
                 //Camera.main.GetComponent<MouseLook>().enabled = true;
-                Screen.lockCursor = true;
+                // TODO: Find out how this works.
```
I haven't even tested this code.

Lithrun replied:
That's within coreeditor, aka the old level editor which we don't need. Commented out all code within this method and added a NotImplementedException in #61

mwnDK1402 replied:
I'll leave this alone and make sure we override this in the merge with #61.

mwnDK1402 commented on Assets/Scripts/FengGameManagerMKII.cs:
```diff
@@ -942,6 +932,7 @@ private void coreeditor()
                         this.selectedObj = transform8.gameObject;
                         //TODO Mouselook
                         //Camera.main.GetComponent<MouseLook>().enabled = false;
+                        Debug.Log("Unexplained 'lockCursor = true' reached");
```
I try to put messages like these so I'll see when the code is reached.

Lithrun replied:
Also don't need the Old Level Editor anymore (See above comment)

mwnDK1402 commented on Assets/Scripts/UI/CameraMode.cs:
```diff
@@ -0,0 +1,6 @@
+﻿public enum CameraMode
```
Maybe add one for player titans? That one shouldn't be saved as a preferred camera mode, though, so maybe not.

Lithrun replied:
Player Titans can have a TPS, Original, WOW as well. Identical to HERO (See previous comment at PlayerTitan.cs)

mwnDK1402 commented on Assets/Scripts/UI/CursorMode.cs:
```diff
@@ -0,0 +1,5 @@
+﻿public enum CursorMode
+{
+    Menu,
+    Loading
```
Should probably change Loading -> Invisible. But in that case, maybe also Menu -> Visible. Thoughts?
@Lithrun

Lithrun replied:
Visible / Invisible would be preferred if you want only want to indicate that the cursor is visible, or invisible

mwnDK1402 replied:
I'll keep the implementation separate from the interface. Keeping it as `Menu` and `Loading`.

mwnDK1402 commented on Assets/Scripts/UI/GameCursor.cs:
```diff
@@ -0,0 +1,153 @@
+    private static void LoadPlayerPrefs()
+    {
+        var savedValue = PlayerPrefs.GetInt(CameraModeKey, (int)CameraMode);
+        if (Enum.IsDefined(typeof(CameraMode), savedValue))
```
I added this, because I had changed the type of enum that was stored in PlayerPrefs... So the old value was no longer defined by the new enum.

Lithrun replied:
So if this only occurs with an earlier version of 164, then we don't need this code within our branch

mwnDK1402 replied:
> So if this only occurs with an earlier version of 164, then we don't need this code within our branch

It's fixing a potentially breaking change should we ever change the enum. Let's keep it; it doesn't do any harm.

mwnDK1402 commented on Assets/Scripts/UI/GameCursor.cs:
```diff
@@ -0,0 +1,153 @@
+
+    private void OnLevelWasLoaded()
+    {
+        if (MenuManager.IsMenuOpen)
```
This is the transition from Loading -> Menu.

mwnDK1402 commented on Assets/Scripts/UI/GameCursor.cs:
```diff
@@ -0,0 +1,153 @@
+
+    private void OnMenuOpened()
+    {
+        CursorMode = CursorMode.Menu;
```
I had put `ApplyCursorMode` here, but that would reapply `Loading` if a menu opens after having loaded a level. Not that it matters, though, since this is only run when all menus are closed, and a menu is opened again. And while loading a level, it always has a menu open, so `OnMenuOpened` is not called when a level finishes loading.

mwnDK1402 commented on Assets/Scripts/UI/InGame/InGameMenu.cs:
```diff
@@ -5,45 +5,44 @@ namespace Assets.Scripts.UI.InGame
         public GraphicSettingMenu GraphicSettingsMenu;
-        public void ShowGameSettingsMenu()
+
+        // Used by Button.
```
I think comments like these are useful, since I was about to make this `private`, thinking it was a Unity message.

mwnDK1402 commented on Assets/Scripts/UI/InGame/RegistrationCounter.cs:
```diff
@@ -0,0 +1,30 @@
+﻿using UnityEngine;
+
+public sealed class RegistrationCounter
```
I like to make my classes sealed when they weren't designed to be extended. If we wanted this to be extendable, we would add more `protected` and `virtual` members. Let me know which is preferred.
@Lithrun

Lithrun replied:
Sealed is fine, however, I found the naming of this class quite strange. "RegistrationCounter", so, we are Registering a Counter. The events also have strange names. It sounds like that you are trying to abstract counters, which perhaps future classes could also use? But if so, then why? Sounds like over-engineering to me.

mwnDK1402 commented on Assets/Scripts/UI/InGame/SpawnMenu.cs:
```diff
@@ -87,10 +87,14 @@ private static bool isPlayerAllDead2()
         {
-            Cursor.visible = true;
-            Screen.lockCursor = false;
+            MenuManager.RegisterOpened();
```
I'm not sure about repeating this in all menus. I like that I could put it into `UiNavigationElement`. Maybe we could avoid this code duplication in menus.

Lithrun replied:
Sounds good

mwnDK1402 commented on Assets/Scripts/UI/Menu/MainMenu.cs:
```diff
@@ -1,14 +1,7 @@
-        private void OnEnable()
-        {
-            Cursor.visible = true;
-            Cursor.lockState = CursorLockMode.Confined; 
```
Reminds me that we should try using `Confined` in Original and WOW. It caused a bug earlier, maybe not anymore?

Lithrun replied:
So investigate and confirm if there's still a bug or not

mwnDK1402 replied:
It still caused a bug, but I just fixed that.

mwnDK1402 commented on Assets/Scripts/UI/Menu/MenuManager.cs:
```diff
@@ -0,0 +1,28 @@
+﻿public static class MenuManager
```
I'm quite happy with this class. I managed to create a class specifically for our implementation of the counter (`RegistrationCounter`), but I also wrapped it in super readable interface (`MenuManager`).

mwnDK1402 reviewed (commented):
I'm satisfied with the changes.

Lithrun commented on Assets/Scripts/FengGameManagerMKII.cs:
```diff
@@ -3512,11 +3456,11 @@ public IEnumerator reloadSky()
-        Screen.lockCursor = !Screen.lockCursor;
-        Screen.lockCursor = !Screen.lockCursor;
+
+        // TODO: Investigate whether this is necessary.
```
So investigate it. Do we need it or not?

mwnDK1402 replied:
We don't. 🥳

Lithrun commented on Assets/Scripts/FengGameManagerMKII.cs:
```diff
@@ -4439,7 +4363,7 @@ public void SpawnPlayerTitan()
         }
         this.myLastHero = id.ToUpper();
-        var playerTitan = PhotonNetwork.Instantiate("PlayerTitan", position, new Quaternion(), 0).GetComponent<PlayerTitan>();
+        PlayerTitan playerTitan = PhotonNetwork.Instantiate("PlayerTitan", position, new Quaternion(), 0).GetComponent<PlayerTitan>();
```
Why change this var to PlayerTitan? Seems quite random.

mwnDK1402 replied:
I couldn't see what type it was, so I just quickly changed it. o.o
The `.GetComponent<PlayerTitan>()` was all the way off my screen.

Lithrun commented on Assets/Scripts/UI/Camera/IN_GAME_MAIN_CAMERA.cs:
```diff
@@ -78,136 +77,36 @@ public class IN_GAME_MAIN_CAMERA : MonoBehaviour
-    }
-
-    private void camareMovement()
+    public enum RotationAxes
```
Public enums should be moved in their own file. Multiple classes use it, so it's confusing that it's located within one of many classes. If it's only used within this class, then just make it private as private enums are fine.

mwnDK1402 replied:
I didn't create this enum, I just reorganized it. I'll move it to its own file.

mwnDK1402 replied:
It is unused, so I deleted it.

Lithrun commented on Assets/Scripts/UI/Camera/IN_GAME_MAIN_CAMERA.cs:
```diff
@@ -724,12 +486,13 @@ public void update2()
         {
-            Cursor.visible = true;
-            Screen.lockCursor = false;
+            // TODO: Investigate whether this is necessary.
```
So investigate it

mwnDK1402 replied:
I'm going to assume it doesn't matter since we haven't discovered any bugs caused by it so far.

Lithrun commented on Assets/Scripts/UI/Camera/IN_GAME_MAIN_CAMERA.cs:
```diff
@@ -840,29 +599,14 @@ public void update2()
-                    }
+                    GameCursor.Cycle();
+
+                    // TODO: Figure out what this does.
```
So what does it do? Might be important to this issue

mwnDK1402 replied:
There are no other references to this field, so I'm deleting it.

Lithrun commented on Assets/Scripts/UI/Camera/IN_GAME_MAIN_CAMERA.cs:
```diff
@@ -840,29 +599,14 @@ public void update2()
                 }
                 if (this.inputManager.isInputDown[InputCode.hideCursor])
                 {
+                    // TODO: Incorporate this in CursorManagement.
```
So then incorporate it with CursorManagement

mwnDK1402 replied:
It is done.

Lithrun commented on Assets/Scripts/UI/Camera/IN_GAME_MAIN_CAMERA.cs:
```diff
@@ -969,11 +713,262 @@ public void update2()
+        transform3.position -= (Vector3)(((base.transform.forward * this.distance) * this.distanceMulti) * this.distanceOffsetMulti);
+    }
+
+    private void CreateMinimap()
```
Method does nothing, so can be removed?

mwnDK1402 replied:
> Method does nothing, so can be removed?

I don't know anything about this. I'm guessing this is still here so we know how it was originally implemented?

mwnDK1402 replied:
I removed it, along with many other unused members.

Lithrun commented on Assets/Scripts/UI/Camera/IN_GAME_MAIN_CAMERA.cs:
```diff
@@ -969,11 +713,262 @@ public void update2()
+        Vector3 position = this.main_object.transform.position;
+        foreach (GameObject obj3 in objArray)
+        {
+            Vector3 vector2 = obj3.transform.Find("Amarture/Core/Controller_Body/hip/spine/chest/neck").position - position;
```
How often is this method called? Please use MindlessTitan.TitanBody.Neck for performance reasons

mwnDK1402 replied:
It's called once for each titan in the level, every time you try to lock on to a titan.

mwnDK1402 replied:
> Please use MindlessTitan.TitanBody.Neck for performance reasons

This requires using `FindGameObjectsWithTag` on top of `GetComponent<MindlessTitan>`. I think a future optimization should be to have some sort of index for all titans. So you can just say `foreach(var titan in TitanManager.AllTitans)`. I'm not convinced using `MindlessTitan` right now will make much of a difference, so I'm adding a TODO for a future performance optimization.

Lithrun commented on Assets/Scripts/UI/Camera/IN_GAME_MAIN_CAMERA.cs:
```diff
@@ -969,11 +713,262 @@ public void update2()
+        return obj2;
+    }
+
+    private int getReverse()
```
Methods are always PascalCase

mwnDK1402 replied:
Should I refactor the whole class?

Lithrun commented on Assets/Scripts/UI/Menu/CreateRoom.cs:
```diff
@@ -43,7 +42,11 @@ public void Start()
         public void Create()
         {
-            //TODO: This will fail when multiple input fields are present.
+            // TODO: Find a better place to put this.
```
Find a better place for it, so why isn't a better place found yet?

mwnDK1402 replied:
I'd like some help, because I'm not very familiar with the code.
Where is the code for loading a level?
Just in general, no matter whether you're creating a game or joining one.
I want to set the cursor mode for loading a level, so it has to run before the level starts loading.

Lithrun replied:
The level itself is loaded via OnJoinedRoom inside the GameManager, with PhotonNetwork.LoadLevel(Level.SceneName);

mwnDK1402 replied:
Done.

Lithrun reviewed (changes requested):
See comments, additionally:
- [x] Please remove any Debug.Log as we don't need that within our development branch.
- [x] Merge with latest development to include the EditorConfig and use Ctrl K + D to apply the correct styling conventions

mwnDK1402 committed:
Remove unneeded hack

I was unable to reproduce the bug that the hack was intended to fix.

mwnDK1402 committed:
Apply styling convention and refactor

mwnDK1402 committed:
Remove hack forcing TPS for player titans

mwnDK1402 committed:
Remove unused enum and field

mwnDK1402 committed:
Remove unnecessary TODO

It is assumed the removed code is an unnecessary transition to CursorMode.Menu when game is stopped.

mwnDK1402 committed:
Remove unnecessary field

mwnDK1402 committed:
Implement ToggleCursor behaviour

Rename InputCode.hideCursor -> toggleCursor.
Implement behaviour in GameCursor through ForceFreeCursor property.

mwnDK1402 committed:
Apply styling convention

mwnDK1402 committed:
Remove unused members

mwnDK1402 committed:
Enable lock-on

This required setting an input key for lock-on, uncommenting references to "locker" GameObject, which doesn't exist, and adding a TODO for optimizing access to the titan's positions, which is required to find the nearest titan to lock on to.

mwnDK1402 committed:
Set WOW and Original lockState to Confined

Makes cursor confined to the window in the standalone build.

mwnDK1402 committed:
Refactor SetPreferredCameraMode to actually use its parameter

mwnDK1402 committed:
Remove unnecessary commented out code

mwnDK1402 committed:
Merge with development

mwnDK1402 committed:
Remove Debug.Log calls

mwnDK1402 committed:
Move CursorMode.Loading to OnJoinedRoom

CursorMode.Loading will now not only be set when a room is created, it is also set when you join a room.

mwnDK1402 committed:
Make cursor visible while loading

mwnDK1402 committed:
Merge with development

mwnDK1402 committed:
Merge remote-tracking branch 'origin/development' into #164-cursor-overhaul

Lithrun commented:
No issues reported for at least 5 days. PR can be finished.
